src/app/GitUI/LeftPanel/FavoritesTree.cs

```diff
@@ -0,0 +1,168 @@
+        {
+            _aheadBehindDataProvider = aheadBehindDataProvider;
+            _revisionGridInfo = revisionGridInfo;
+            _refUiCommand = uiCommands;
+
+            if (refsSource is GitModuleControl { Module.WorkingDirGitDir: not null } gitModuleControl)
+            {
+                _cachedFavorites.Location = gitModuleControl.Module.WorkingDirGitDir;
+            }
+        }
+
+        public void Add(NodeBase node)
+        {
+            if (node is BaseRevisionNode baseRevisionNode && baseRevisionNode.ObjectId != null)
+            {
+                _cachedFavorites.Add(baseRevisionNode.ObjectId.ToString());
+            }
+
+            Refresh(new FilteredGitRefsProvider(UICommands.Module).GetRefs, true);
+        }
+
+        public void Remove(NodeBase node)
+        {
+            if (node is BaseRevisionNode baseRevisionNode && baseRevisionNode.ObjectId != null)
+            {
+                _cachedFavorites.Remove(baseRevisionNode.ObjectId.ToString());
+            }
+
+            Refresh(new FilteredGitRefsProvider(UICommands.Module).GetRefs, true);
+        }
+
+        protected override Nodes FillTree(IReadOnlyList<IGitRef> branches, CancellationToken token)
+        {
+            FavoriteNode remoteNode = new(this, TranslatedStrings.Remotes, nameof(Images.BranchRemoteRoot)) { Visible = true };
+            FavoriteNode localNode = new(this, TranslatedStrings.Local, nameof(Images.BranchLocalRoot)) { Visible = true };
+
+            Dictionary<string, BaseRevisionNode> pathToNodesRemote = [];
+            Dictionary<string, BaseRevisionNode> pathToNodeLocal = [];
+            IDictionary<string, AheadBehindData> aheadBehindDataLocal = _aheadBehindDataProvider?.GetData();
+
+            foreach (IGitRef branch in branches)
+            {
+                token.ThrowIfCancellationRequested();
+
+                if (branch.IsRemote)
+                {
+                    CreateRemoteBranchTree(branch, pathToNodesRemote, remoteNode.Nodes, aheadBehindDataLocal);
+                }
+                else if (branch.IsHead)
+                {
+                    CreateLocalBranchTree(branch, pathToNodeLocal, localNode.Nodes, aheadBehindDataLocal);
+                }
+            }
+
+            _cachedFavorites.Favorites.ToList().Except(branches.Select(p => p.ObjectId?.ToString())).ForEach(p => _cachedFavorites.Remove(p));
+
+            Nodes nodes = new(this);
+            nodes.AddNode(localNode);
+            nodes.AddNode(remoteNode);
+
+            return nodes;
+        }
+
+        protected override void PostFillTreeViewNode(bool firstTime)
+        {
+            if (firstTime)
+            {
+                TreeViewNode.Collapse();
+            }
+        }
+
+        private void CreateLocalBranchTree(IGitRef branch, Dictionary<string, BaseRevisionNode> pathToNodes, Nodes nodes, IDictionary<string, AheadBehindData>? aheadBehindData)
+        {
+            Validates.NotNull(branch.ObjectId);
```

Comment:
Before calling this function, the caller should filter out special branches which lack an `ObjectId` for whatever reason.